File: typescript/packages/core/src/classes/WalletClientBase.ts
import { z } from "zod";
import type { Chain } from "../types/Chain";
import type { Token } from "../types/Token";
import { type ToolBase, createTool } from "./ToolBase";

export type Signature = {
    signature: string;
};

export type Balance = {
    decimals: number;
    symbol: string;
    name: string;
    value: string;
    inBaseUnits: string;
};

export abstract class WalletClientBase {
    abstract getAddress(): string;
    abstract getChain(): Chain;
    abstract signMessage(message: string): Promise<Signature>;
    abstract balanceOf(address: string, tokenAddress?: string): Promise<Balance>;

    getTokenInfoByTicker(ticker: string): Promise<Token> {
        throw new Error("getTokenInfoByTicker is not implemented for this wallet client");
    }

    getCoreTools(): ToolBase[] {
        return [
            createTool(
                {
                    name: "get_address",
                    description: "Get the address of the wallet",
                    parameters: z.object({}),
                },
                () => this.getAddress(),
            ),
            createTool(
                {
                    name: "get_chain",
                    description: "Get the chain of the wallet",
                    parameters: z.object({}),
                },
                () => this.getChain(),
            ),
            createTool(
                {
                    name: "get_balance",
                    description:
                        "Get the balance of the wallet for the native token or a specific token by passing its address.",
                    parameters: z.object({
                        address: z.string(),
                        tokenAddress: z.string().optional(),
                    }),
                },
                (parameters) => this.balanceOf(parameters.address, parameters.tokenAddress),
            ),
            createTool(
                {
                    name: "sign_message",
                    description: "Sign a message with the wallet",
                    parameters: z.object({ message: z.string() }),
                },
                (parameters) => this.signMessage(parameters.message),
            ),
        ];
    }
}

File: python/src/goat-sdk/goat/classes/wallet_client_base.py
from abc import ABC, abstractmethod
from typing import List, Optional, TypedDict
from pydantic import BaseModel, Field

from goat.classes.tool_base import ToolBase, create_tool
from goat.types.token import Token
from goat.types.chain import Chain

class EmptyParams(BaseModel):
    pass


class BalanceParams(BaseModel):
    address: str = Field(description="The wallet address to check the balance of")
    tokenAddress: Optional[str] = Field(description="The token address to check the balance of", default=None)


class Signature(TypedDict):
    signature: str


class Balance(TypedDict):
    decimals: int
    symbol: str
    name: str
    value: str
    in_base_units: str


class WalletClientBase(ABC):
    @abstractmethod
    def get_address(self) -> str:
        pass

    @abstractmethod
    def get_chain(self) -> Chain:
        pass

    @abstractmethod
    def sign_message(self, message: str) -> Signature:
        pass

    @abstractmethod
    def balance_of(self, address: str, token_address: Optional[str] = None) -> Balance:
        pass

    def get_token_info_by_ticker(self, ticker: str) -> Token:
        raise NotImplementedError("get_token_info_by_ticker is not implemented for this wallet client")

    def get_core_tools(self) -> List[ToolBase]:
        return [
            create_tool(
                {
                    "name": "get_address",
                    "description": "Get the address of the wallet",
                    "parameters": EmptyParams
                },
                lambda _: self.get_address(),
            ),
            create_tool(
                {
                    "name": "get_chain",
                    "description": "Get the chain of the wallet",
                    "parameters": EmptyParams
                },
                lambda _: self.get_chain(),
            ),
            create_tool(
                {
                    "name": "get_balance",
                    "description": "Get the balance of the wallet",
                    "parameters": BalanceParams
                },
                lambda parameters: self.balance_of(parameters["address"], parameters.get("tokenAddress")),
            ),
        ]
